### PR TITLE
Enable FindBugs XML reports in the plugin

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] Reconfigure pom.xml
- [x] Enable FindBugs on https://jenkins.ci.cloudbees.com/job/plugins/job/ssh-slaves-plugin/
- [x] Fix important issues

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@reviewbybees
Cleaned up all issues discovered by FindBugs

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 1]
```diff
@@ -787,8 +806,7 @@ protected String resolveJava(SlaveComputer computer, TaskListener listener) thro
                 try {
                     tried.add(javaCommand);
                     String java = checkJavaVersion(listener, javaCommand);
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: just return it?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 2]
```diff
@@ -852,7 +871,14 @@ private EnvVars getEnvVars(DescribableList<NodeProperty<?>, NodePropertyDescript
     private void verifyNoHeaderJunk(TaskListener listener) throws IOException, InterruptedException {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
         connection.exec("true",baos);
-        String s = baos.toString();
+        final String s;
+        //TODO: Seems we need to retrieve the encoding from the connection destination
+        try {
+            s = baos.toString(Charset.defaultCharset().name());
+        } catch (UnsupportedEncodingException ex) { // Should not happen
+            throw new IOException("Default encoding is unsupported");
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: `, ex`?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 3]
```diff
@@ -883,8 +909,14 @@ private String attemptToInstallJDK(TaskListener listener, String workingDirector
...
         //        (this one is from MKS)
-
-        String uname = unameOutput.toString();
+        
+        //TODO: Seems we need to retrieve the encoding from the connection destination
+        final String uname;
+        try {
+            uname = unameOutput.toString(Charset.defaultCharset().name());
+        } catch (UnsupportedEncodingException ex) { // Should not happen
+            throw new IOException("Default encoding is unsupported");
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: `, ex`?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 4]
```diff
@@ -1071,13 +1103,16 @@ protected void reportEnvironment(TaskListener listener) throws IOException, Inte
...
+    @NonNull
     private String checkJavaVersion(TaskListener listener, String javaCommand) throws IOException, InterruptedException {
         listener.getLogger().println(Messages.SSHLauncher_CheckingDefaultJava(getTimestamp(),javaCommand));
         StringWriter output = new StringWriter();   // record output from Java
 
         ByteArrayOutputStream out = new ByteArrayOutputStream();
         connection.exec(javaCommand + " "+getJvmOptions() + " -version",out);
-        BufferedReader r = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(out.toByteArray())));
+        //TODO: Seems we need to retrieve the encoding from the connection destination
+        BufferedReader r = new BufferedReader(new InputStreamReader(
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: wouldn't it cause resource leak?

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 5]
```diff
@@ -457,8 +460,9 @@ public SSHLauncher(String host, int port, StandardUsernameCredentials credential
         this.prefixStartSlaveCmd = fixEmpty(prefixStartSlaveCmd);
         this.suffixStartSlaveCmd = fixEmpty(suffixStartSlaveCmd);
         this.launchTimeoutSeconds = launchTimeoutSeconds == null || launchTimeoutSeconds <= 0 ? null : launchTimeoutSeconds;
-        this.maxNumRetries = maxNumRetries == null || maxNumRetries <= 0 ? 0 : maxNumRetries;
-        this.retryWaitTime = retryWaitTime == null || retryWaitTime <= 0 ? 0 : retryWaitTime;
+        final Integer zero = 0;
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: `Integer.valueOf(0)`?

Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: Any constants in the range -128..127 will be subst by the compiler anyway

File: src/main/java/hudson/plugins/sshslaves/SSHConnector.java [hunk 6]
```diff
@@ -259,8 +261,9 @@ public SSHConnector(int port, StandardUsernameCredentials credentials, String us
         this.prefixStartSlaveCmd = fixEmpty(prefixStartSlaveCmd);
         this.suffixStartSlaveCmd = fixEmpty(suffixStartSlaveCmd);
         this.launchTimeoutSeconds = launchTimeoutSeconds == null || launchTimeoutSeconds <= 0 ? null : launchTimeoutSeconds;
-        this.maxNumRetries = maxNumRetries == null || maxNumRetries <= 0 ? 0 : maxNumRetries;
-        this.retryWaitTime = retryWaitTime == null || retryWaitTime <= 0 ? 0 : retryWaitTime;
+        final Integer zero = 0;
+        this.maxNumRetries = maxNumRetries == null || maxNumRetries <= 0 ? zero : maxNumRetries;
+        this.retryWaitTime = retryWaitTime == null || retryWaitTime <= 0 ? zero : retryWaitTime;
```
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHConnector.java]: I would have gone with `retryWaitTime == null ? 0 : Math.max(0, retryWaitTime)` or `Math.max(0, retryWaitTime == null? 0 : retryWaitTime)` or `retryWaitTime == null || retryWaitTime <= 0 ? 0 : (int)retryWaitTime` adding a local constant seems insane just to stop complaints about covariant returns from an elvis operator

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 4 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: There is no resource leak here, because we don't use any file resources here

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 5 above)
Comment (KostyaSha) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: ah, right. java compiler is also smart as gcc

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 4 above)
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: not with a byte array stream as there are no external resources to leak... but it could cause complaints of a resource leak...

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 7]
```diff
@@ -1216,7 +1252,7 @@ public synchronized void afterDisconnect(SlaveComputer slaveComputer, TaskListen
                     if (sftpClient == null) {// system without SFTP
                         try {
                             connection.exec("rm " + fileName, listener.getLogger());
-                        } catch (Exception x) {
+                        } catch (Throwable x) {
```
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: What about letting `Error` fall through. Seems like a change above and beyond findbugs

File: pom.xml [hunk 8]
```diff
@@ -14,6 +14,11 @@
   <name>Jenkins SSH Slaves plugin</name>
   <url>http://wiki.jenkins-ci.org/display/JENKINS/SSH+Slaves+plugin</url>
 
+  <properties>
+    <findbugs-maven-plugin.version>3.0.1</findbugs-maven-plugin.version>
```
Comment (stephenc) [inline pom.xml]: I *hate* people using properties for a dependency/plugin version that is used once and only once in the pom. Using properties for versions is really not a good plan unless you expressly know exactly what you are doing and why you are doing it.

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 4 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: @stephenc 
It does not cause complains at least in FindBugs 3.0.1

File: pom.xml [hunk 9]
```diff
@@ -14,6 +14,11 @@
...
+    <findbugs-maven-plugin.failOnError>true</findbugs-maven-plugin.failOnError>
```
Comment (stephenc) [inline pom.xml]: The [user property](http://gleclaire.github.io/findbugs-maven-plugin/check-mojo.html#failOnError) is `findbugs.failOnError` do not hijack the expected property and move it to a different name

File: pom.xml [hunk 10]
```diff
@@ -101,6 +112,25 @@
...
         <version>2.8</version>
       </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>findbugs-maven-plugin</artifactId>
+        <version>${findbugs-maven-plugin.version}</version>
+        <configuration>
+          <xmlOutput>true</xmlOutput>
+          <findbugsXmlWithMessages>true</findbugsXmlWithMessages>
+          <failOnError>${findbugs-maven-plugin.failOnError}</failOnError>
```
Comment (stephenc) [inline pom.xml]: Do not hijack the existing user property. If a parent has forced this to `true` or `false` and you want to switch it back, then use `${findbugs.failOnError}` not a random property name that does not fit in with the existing maven plugin documentation

File: pom.xml (hunk 8 above)
Comment (KostyaSha) [inline pom.xml]: it exists in upstream pom

File: src/main/java/hudson/plugins/sshslaves/SSHConnector.java (hunk 6 above)
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHConnector.java]: Actually, wouldn't the correct thing be to change the field type to `int` as that is the default value when `null` and that would remove the boxing/unboxing

File: pom.xml (hunk 8 above)
Comment (stephenc) [inline pom.xml]: Ugh! Yes I forgot that the upstream pom does every kind of wrong under the sun

File: pom.xml (hunk 10 above)
Comment (KostyaSha) [inline pom.xml]: i think @oleg-nenashev followed existing style with controlling jenkins build processes with properties that come from 'pom' project. Handling env properties from profile also makes sense but seems doesn't fit idea with enforcing it.

File: pom.xml (hunk 8 above)
Comment (KostyaSha) [inline pom.xml]: Well, it not so bad imho in comparison to redundant ancient pom project https://github.com/jenkinsci/pom

File: pom.xml (hunk 9 above)
Comment (KostyaSha) [inline pom.xml]: @oleg-nenashev maybe just remove this variable and set true in configuration section?

Comment (KostyaSha) [inline pom.xml]: @stephenc please check maillist i sent my thought about this variables. Maybe you can propose something better.

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java (hunk 7 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: We got into several critical failures in buggy Jenkins extensions due to catching Exceptions and Missing throwables. Probably for this code it makes sense to accept Errors

File: src/main/java/hudson/plugins/sshslaves/SSHConnector.java (hunk 6 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sshslaves/SSHConnector.java]: The field is public => no easy way back to <code>int</code>

Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sshslaves/SSHConnector.java]: @stephenc 
I'll change the order of conditions to <code>this.maxNumRetries = maxNumRetries != null && maxNumRetries > 0 ? maxNumRetries : 0;</code>. It will prevent a pointless boxing-unboxing cycle, because Integer will be taken as a return type of the operator

**Comment by oleg-nenashev** [comment]
@stephenc @KostyaSha 
I've fixed your comments

File: pom.xml (hunk 9 above)
Comment (stephenc) [inline pom.xml]: There is nothing to think about IMHO. the existing property used to control this configuration option is called `findbugs.failOnError` as per the plugins own [documentation](http://gleclaire.github.io/findbugs-maven-plugin/check-mojo.html#failOnError). 

I might have personal complaints about that specific plugin having picked a crummy name for the property... but that does not change the fact that the convention of all maven users is to use the property described in the plugin's documentation.

So the name should be `findbugs.failOnError` (even though I do not like that specific property name on so many levels... those complaints are to be addressed with the findbugs plugin maintainers)

**Comment by stephenc** [comment]
:bug: due to not following maven conventions for `findbugs.failOnError` property name

File: pom.xml (hunk 9 above)
Comment (KostyaSha) [inline pom.xml]: +1 for renaming or just hardcode it in configuration section.

Comment (oleg-nenashev) [inline pom.xml]: Agreed

**Comment by stephenc** [comment]
:bee:

**Comment by KostyaSha** [comment]
:+1:

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 11]
```diff
@@ -1012,10 +1042,10 @@ private void copySlaveJar(TaskListener listener, String workingDirectory) throws
                         os.close();
                     }
                     listener.getLogger().println(Messages.SSHLauncher_CopiedXXXBytes(getTimestamp(), slaveJar.length));
-                } catch (Exception e) {
+                } catch (Throwable e) {
```
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: You should let `Error` passthrough as is. Escalating to `Throwable` without a passthrough for `Error` is wrong

Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: Also changing from Exception to Throwable should be a separate PR IMHO

File: src/main/java/hudson/plugins/sshslaves/SSHLauncher.java [hunk 12]
```diff
@@ -1216,8 +1250,11 @@ public synchronized void afterDisconnect(SlaveComputer slaveComputer, TaskListen
                     if (sftpClient == null) {// system without SFTP
                         try {
                             connection.exec("rm " + fileName, listener.getLogger());
-                        } catch (Exception x) {
+                        } catch (Throwable x) {
                             x.printStackTrace(listener.error(Messages.SSHLauncher_ErrorDeletingFile(getTimestamp())));
+                            if (x instanceof Error) {
```
Comment (stephenc) [inline src/main/java/hudson/plugins/sshslaves/SSHLauncher.java]: Ugh! nope... that's not the way to do it

```
} catch (Error x) { 
  throw x;
} catch (Throwable x) {
```
is the way to do it

**Comment by stephenc** [comment]
:bug: exception mis-handling of `Error`s

**Comment by oleg-nenashev** [comment]
@stephenc 
Fixed the exceptions

**Comment by stephenc** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
@reviewbybees done
Squashing commits
